File: brew_wizard_server/app/models/yeast.rb
class Yeast < ApplicationRecord
  belongs_to :user
  has_many :recipe_yeast, dependent: :destroy
  has_many :recipes, :through => :recipe_yeast

  has_many :style_yeasts, dependent: :destroy
  has_many :styles, :through => :style_yeasts

  has_many :yeast_relations, class_name: 'YeastRelation', foreign_key: :yeast_id
  has_many :inverse_yeast_relations, class_name: 'YeastRelation', foreign_key: :yeast_relation_id
  has_many :yeast_substitutes, through: :yeast_relations
  has_many :inverse_yeast_substitutes, through: :inverse_yeast_relations

  validates :name, presence: true, length: { maximum: 40 }
  validates :lab, length: { maximum: 40 }
  validates :product_id, length: { maximum: 10 }
  validates :yeast_type, presence: true, length: { maximum: 10 }
  validates :form, length: { maximum: 10 }
  validates :cell_count, presence: true, numericality: { greater_than: 0}
  validates :min_attenuation, presence: true, numericality: { greater_than_or_equal_to: 0, less_than_or_equal_to: 100}
  validates :max_attenuation, presence: true, numericality: { greater_than_or_equal_to: 0, less_than_or_equal_to: 100}
  validates :flocculation, presence: true, numericality: { greater_than_or_equal_to: 0, less_than_or_equal_to: 5}
  validates :min_temperature, presence: true, numericality: { greater_than_or_equal_to: 0, less_than_or_equal_to: 212}
  validates :max_temperature, presence: true, numericality: { greater_than_or_equal_to: 0, less_than_or_equal_to: 212}
  validates :description, length: { maximum: 2000 }

  accepts_nested_attributes_for :style_yeasts
  accepts_nested_attributes_for :yeast_relations

  def recipe_count
    return self.recipes.count
  end
end

File: brew_wizard_ui/src/app/yeast/yeast.ts
import { IStyle } from './../style/style';

export interface IYeast {
  id: number;
  user_id: number;
  name: string;
  global: boolean;
  lab: string;
  product_id: string;
  yeast_type: string;
  form: string;
  flocculation: number;
  min_attenuation: number;
  max_attenuation: number;
  min_temperature: number;
  max_temperature: number;
  cell_count: number;
  description: string;
  recipe_count: number;
  style_yeasts: { id: number, style_id: number, yeast_id: number }[];
  yeast_relations: { id: number, yeast_id: number, yeast_relation_id: number }[];
}
